### DISCUSSION TITLE
${GITHUB_SHA} isn't parsed

### DISCUSSION THREAD

**Discussion by Andre601** [community/community#25356] [Actions]
<p>I have a GitHub Action set up, which automatically pushes changes from a documentation to the <code>gh-pages</code> branch.<br>
In that action do I use <code>${GITHUB_SHA:0:7}</code> to display the first 7 characters of the commit SHA. This works for the first job, but for the second one, which sends a notification to a Discord channel using webhooks, is it not working.<br>
The placeholder remains as is without any changes while other placeholders like <code>${{ github.event.compare }}</code> gets translated properly.</p>
<p>Could an issue be, that <code>${GITHUB_SHA:0:7}</code> is surrounded by back ticks (<code>`</code>)?<br>
The code in general looks like this: <code>[`${GITHUB_SHA:0:7}`](${{ github.event.compare }})</code></p>
<p>Or would I need to first checkout the job again in order to set the env variable first?</p>

**Comment by Yanjingzhu** [comment]
<p>Hi <a class="mention" href="/Andre601">@Andre601</a>,<br>
Based on my test, when using back-ticks (`),  ${GITHUB_SHA:0:7} could not get the expected value.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/722b9419e034115cbc9880451d490c5c1071ee75.png" data-download-href="/uploads/short-url/ghZNucu0JTjCw6gGJfI5calt7Qp.png?dl=1" title="sha_back" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181090591-b1d9407d-1204-48c7-bb49-cf6415277647.png" alt="sha_back" data-base62-sha1="ghZNucu0JTjCw6gGJfI5calt7Qp" width="690" height="169" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/722b9419e034115cbc9880451d490c5c1071ee75_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">sha_back</span><span class="informations">798×196 13.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
Remove back-ticks could work fine.<br>
<div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/46861da3dbc7d2ab3fbe4e1a27722764a523b583.png" data-download-href="/uploads/short-url/a3SLpnPGbIYhqkc2ZXJnOqkNv2j.png?dl=1" title="sha07" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181090594-246095bb-fdc3-4e55-852f-2da09556a1a4.png" alt="sha07" data-base62-sha1="a3SLpnPGbIYhqkc2ZXJnOqkNv2j" width="690" height="167" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/46861da3dbc7d2ab3fbe4e1a27722764a523b583_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">sha07</span><span class="informations">793×192 8.56 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
If remove back-ticks could not work in your scenario. Could you share your whole workflow yml here?<p></p>

**Comment by Andre601** [comment]
<p>It does work when removing the backticks, but my point is, that I want to have those backticks to format the sha.<br>
Is there any alternative to display the first 7 characters of the commit sha, which would work with the backticks provided?</p>
<p>If not would I hope that a GitHub dev/employe could look into fixing this rather annoying issue as I see no reason why the variable shouldn’t be translated here.</p>

**Comment by airtower-luna** [comment]
<p>The problem is that backticks are part of shell syntax, and bash will try to execute the string between the backticks and substitute the output. Hence the “1503d45: command not found” in the example with backticks posted by <a class="mention" href="/yanjingzhu">@yanjingzhu</a>.</p>
<p>Try escaping the backticks by placing backslashes in front of them, like this:</p>
<pre><code class="lang-bash">echo "[\`${GITHUB_SHA:0:7}\`](${{ github.event.compare }})"
</code></pre>

**Comment by Andre601** [comment]
<p>Thanks for the suggestion.</p>
<p>Unfortunately does this not work, because It causes an error:<br>
<code>##[error]Unexpected token ` in JSON at position 264</code></p>
<p>So escaping it won’t really work it seems.</p>
<p>Here is my full deploy.yml and the particular part I use to send notifications.<br>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/purrbot-site/Docs/blob/70f23d54c6154591a8846502782d7b7be5286569/.github/workflows/deploy.yml#L33-L68" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/purrbot-site/Docs/blob/70f23d54c6154591a8846502782d7b7be5286569/.github/workflows/deploy.yml#L33-L68" target="_blank" rel="nofollow noopener">purrbot-site/Docs/blob/70f23d54c6154591a8846502782d7b7be5286569/.github/workflows/deploy.yml#L33-L68</a></h4>
<pre class="onebox"><code class="lang-yml"><ol class="start lines" start="33" style="counter-reset: li-counter 32 ;">
<li>sendMsg:</li>
<li>  needs: [build]</li>
<li>  if: success()</li>
<li>  runs-on: ubuntu-latest</li>
<li>  steps:</li>
<li>  - name: Send Discord Notification</li>
<li>    uses: Sniddl/discord-commits@v1.2</li>
<li>    with:</li>
<li>      webhook: ${{ secrets.DISCORD_WEBHOOK }}</li>
<li>      embed: |</li>
<li>        {</li>
<li>          "color": 65280,</li>
<li>          "title": "Documentation updated!",</li>
<li>          "url": "https://github.com/purrbot-site/Docs",</li>
<li>          "description": "${{ github.event.commits[0].message }}",</li>
<li>          "footer": {</li>
<li>            "text": "Page built using MkDocs"</li>
<li>          },</li>
<li>          "fields": [</li>
<li>            {</li>
</ol></code></pre>

  This file has been truncated. <a href="https://github.com/purrbot-site/Docs/blob/70f23d54c6154591a8846502782d7b7be5286569/.github/workflows/deploy.yml#L33-L68" target="_blank" rel="nofollow noopener">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
</p>

**Comment by airtower-luna** [comment]
<p>Well, a Bash escape sequence isn’t necessarily going to work in JSON. In JSON</p>
<pre><code class="lang-auto">\`
</code></pre>
<p>isn’t a valid escape sequence. :wink: You should be able to use a literal backtick, but if you really need to escape it you can use a Unicode escape: <code>"\u0060"</code></p>
<p>However, after fixing that you’ll have another problem: The shell variable substitution <code>${GITHUB_SHA:0:7}</code> is probably not going to work outside a shell context. I don’t see anything in the <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#functions" rel="nofollow noopener">function documentation</a> that would cut the SHA string the way you need it, but you could make a separate shell script that sets it as an environment variable or output.</p>

**Comment by Andre601** [comment]
<p>I now tried it without any backticks and even that doesn’t seem to make it translate into the commit-hash.</p>
<p>Does it perhaps require a checkout first, or would the checkout from the previous job work as-well?<br>
Also, <code>${GITHUB_SHA:0:7}</code> does work since I have it used in the job before that one and there it successfully translates into the first 7 characters of the commit sha used.</p>
<p>I really want nothing more than using the commit sha inside a Discord webhook while having it in a inline code block, but this entire shell thing messes that up, which is nothing less than annoying.</p>

**Comment by Andre601** [comment, marked as answer]
<p>I now moved to use <code>${{github.event.after}}</code> since those kind of placeholders seem to work no matter what and while it isn’t the result I want (Only first 7 characters of a commit) is it more than good enough for it.</p>
